File: 61-rotate-list/61-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(k==0 || head==null || head.next==null)
            return head;
        
        // find length of list
        int length=1;
        ListNode curr=head;
        
        while(curr.next!=null)
        {
            length++;
            curr=curr.next;
        }
        
        curr.next=head; // forming a cycle
        k=k%length; // to avoid unecessary rotations
        k=length-k;
        
        while(k-- > 0)
            curr=curr.next;
        
        // break the node
        head=curr.next;
        curr.next=null;
        
        return head;
    }
}

File: 61-rotate-list/61-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head==NULL or k==0)
            return head;
        
        ListNode* temp=head;
        
        int len=1;
        while(temp->next!=NULL)
        {
            len++;
            temp=temp->next;
        }
        
        temp->next=head;
        k%=len;
        k=len-k;
        
        while(k-- >0)
            temp=temp->next;
        
        head=temp->next;
        temp->next=NULL;
        
        return head;
        
    }
};